### PR TITLE
Bug 1865086 - More FxA error reporting

### PR DESCRIPTION
**Pull request by bendk**
Include the exception message when reporting
 FxaDeviceConstellationException.EnsureCapabilitiesFailed
 (https://bugzilla.mozilla.org/show_bug.cgi?id=1865086)

 Report restoring an FxA account that's not connected
 (https://bugzilla.mozilla.org/show_bug.cgi?id=1864667)

 Report a breadcrumb if we queue multiple auth recovery attempts at once
 (https://bugzilla.mozilla.org/show_bug.cgi?id=1864994)

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1865086

### CODE REVIEW

File: android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/AccountStorage.kt [hunk 1]
```diff
@@ -190,6 +196,10 @@ internal abstract class AbnormalAccountStorageEvent : Exception() {
      * Account state was expected to be present, but it wasn't.
      */
     internal class UnexpectedlyMissingAccountState : AbnormalAccountStorageEvent()
+    internal class RestoringNonConnectedAccount(val state: FxaRustAuthState) : AbnormalAccountStorageEvent() {
+        override val message: String
+            get() = "AbnormalAccountStorageEvent.RestoringNonConnectedAccount($state)"
+    }
```
Comment (bendk) [inline android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/AccountStorage.kt:202]: @mavduevskiy Is this the right way to get message strings included in the Sentry report?

Comment (mavduevskiy) [reply android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/AccountStorage.kt:202]: It is certainly a way to get it there! Though one could argue that a more canonical approach for Java/Kotlin throwables would be passing a message into the constructor. We also don't really use `state` as far as I could tell, so 

```
internal class RestoringNonConnectedAccount(override val message: String) : AbnormalAccountStorageEvent()
```
used as
```
crashReporter?.submitCaughtException(
                    AbnormalAccountStorageEvent.RestoringNonConnectedAccount(
                        "AbnormalAccountStorageEvent.RestoringNonConnectedAccount($state)",
                    )
                )
```
could be a way to go.

Comment (bendk) [reply android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/AccountStorage.kt]: I swear I saw a message from you about this, but I don't see it anymore.  In any case, I changed this to use pass a `message` parameter to the exception constructor instead.

File: android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt [hunk 2]
```diff
@@ -449,10 +455,16 @@ open class FxaAccountManager(
     internal suspend fun encounteredAuthError(
         operation: String,
         errorCountWithinTheTimeWindow: Int = 1,
-    ) = withContext(coroutineContext) {
-        processQueue(
-            Event.Account.AuthenticationError(operation, errorCountWithinTheTimeWindow),
-        )
+    ) {
+        if (authRecoveryScheduled.getAndSet(true)) {
+            crashReporter?.recordCrashBreadcrumb(Breadcrumb("multiple auth recoveries scheduled at once"))
```
Comment (mavduevskiy) [inline android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt:460]: How do we know that multiple auth recoveries were scheduled at this point?  The `encounteredAuthError` function is the only place where `authRecoveryScheduled` is set to true. Also, should we rename it `authRecoveryScheduled` to `multipleAuthRecoveryScheduled` if it's the intention of the flag?

**Review by mavduevskiy — CHANGES_REQUESTED** [review]
Looking good! A few changes to make sure that we are passing down the message to the stack trace (should go all the way into the Exception constructor). And a question about `authRecoveryScheduled` usage.

**Comment by bendk** [comment]
Thanks for the help with this one.  I implemented all the suggestions, except the auth check one which I tried to improve a different way.

File: android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt [hunk 3]
```diff
@@ -127,6 +129,19 @@ open class FxaAccountManager(
     @Volatile
     private var latestAuthState: String? = null
 
+    // Used to detect multiple auth recovery attempts at once
+    // (https://bugzilla.mozilla.org/show_bug.cgi?id=1864994)
+    private var authRecoveryScheduled = AtomicBoolean(false)
+    private fun willScheduleAuthRecovery() {
```
Comment (mavduevskiy) [inline android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt:135]: This is much cleaner way to organize the code!

**Review by mavduevskiy — APPROVED** [review]
This looks good! Approving.

One thing to consider is the naming of `willScheduleAuthRecovery` funciton, that sounds like it will schedule the recovery, but it actually checks for multiple scheduled recovers and reports to crash reporter if it's true. My two cents, something like `checkForMultipleRequiveryCalls` would be a better fit.

**Comment by bendk** [comment]
Sounds good, I updated the name and added the `needs-landing` tag.

**Pull request merged**